File: wpimath/src/test/native/cpp/trajectory/TrajectoryTransformTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "wpi/math/TestAssertions.hpp"
#include "wpi/math/geometry/Pose2d.hpp"
#include "wpi/math/trajectory/HolonomicSample.hpp"
#include "wpi/math/trajectory/HolonomicTrajectoryGenerator.hpp"
#include "wpi/units/acceleration.hpp"
#include "wpi/units/angle.hpp"
#include "wpi/units/angular_acceleration.hpp"
#include "wpi/units/angular_velocity.hpp"
#include "wpi/units/length.hpp"
#include "wpi/units/time.hpp"
#include "wpi/units/velocity.hpp"

void TestSameShapedTrajectory(
    const std::vector<wpi::math::HolonomicSample>& statesA,
    const std::vector<wpi::math::HolonomicSample>& statesB) {
  for (unsigned int i = 0; i < statesA.size() - 1; i++) {
    auto a1 = statesA[i].pose;
    auto a2 = statesA[i + 1].pose;

    auto b1 = statesB[i].pose;
    auto b2 = statesB[i + 1].pose;

    auto a = a2.RelativeTo(a1);
    auto b = b2.RelativeTo(b1);

    CHECK_NEAR(a.X().value(), b.X().value(), 1E-9);
    CHECK_NEAR(a.Y().value(), b.Y().value(), 1E-9);
    CHECK_NEAR(a.Rotation().Radians().value(), b.Rotation().Radians().value(),
               1E-9);
  }
}

// A rigid transform rotates both the heading and the field-relative
// velocity/acceleration by the same amount, so the heading-relative forward
// scalars (and curvature) are invariant. This would fail if
// TransformBy/RelativeTo rotated the pose but not the velocity/acceleration.
void TestSameForwardScalars(
    const std::vector<wpi::math::HolonomicSample>& statesA,
    const std::vector<wpi::math::HolonomicSample>& statesB) {
  REQUIRE(statesA.size() == statesB.size());
  for (unsigned int i = 0; i < statesA.size(); i++) {
    CHECK_NEAR(statesA[i].velocity.vx.value(), statesB[i].velocity.vx.value(),
               1E-9);
    CHECK_NEAR(statesA[i].velocity.vy.value(), statesB[i].velocity.vy.value(),
               1E-9);
    CHECK_NEAR(statesA[i].velocity.omega.value(),
               statesB[i].velocity.omega.value(), 1E-9);
    CHECK_NEAR(statesA[i].acceleration.ax.value(),
               statesB[i].acceleration.ax.value(), 1E-9);
    CHECK_NEAR(statesA[i].acceleration.ay.value(),
               statesB[i].acceleration.ay.value(), 1E-9);
    CHECK_NEAR(statesA[i].acceleration.alpha.value(),
               statesB[i].acceleration.alpha.value(), 1E-9);
  }
}

TEST_CASE("TrajectoryTransformsTest TransformBy", "[wpimath]") {
  auto trajectory = wpi::math::HolonomicTrajectoryGenerator::Generate(
      wpi::math::Pose2d{}, {}, wpi::math::Pose2d{1_m, 1_m, 90_deg}, 3_mps,
      1_rad_per_s, 3_mps_sq, 1_rad_per_s_sq);

  auto transformedTrajectory = trajectory.TransformBy({{1_m, 2_m}, 30_deg});

  auto firstPose = transformedTrajectory.SampleAt(0_s).pose;

  CHECK_NEAR(firstPose.X().value(), 1.0, 1E-9);
  CHECK_NEAR(firstPose.Y().value(), 2.0, 1E-9);
  CHECK_NEAR(firstPose.Rotation().Degrees().value(), 30.0, 1E-9);

  TestSameShapedTrajectory(trajectory.Samples(),
                           transformedTrajectory.Samples());
  TestSameForwardScalars(trajectory.Samples(), transformedTrajectory.Samples());
}

TEST_CASE("TrajectoryTransformsTest RelativeTo", "[wpimath]") {
  auto trajectory = wpi::math::HolonomicTrajectoryGenerator::Generate(
      wpi::math::Pose2d{1_m, 2_m, 30_deg}, {},
      wpi::math::Pose2d{5_m, 7_m, 90_deg}, 3_mps, 1_rad_per_s, 3_mps_sq,
      1_rad_per_s_sq);

  auto transformedTrajectory = trajectory.RelativeTo({1_m, 2_m, 30_deg});

  auto firstPose = transformedTrajectory.SampleAt(0_s).pose;

  CHECK_NEAR(firstPose.X().value(), 0, 1E-9);
  CHECK_NEAR(firstPose.Y().value(), 0, 1E-9);
  CHECK_NEAR(firstPose.Rotation().Degrees().value(), 0, 1E-9);

  TestSameShapedTrajectory(trajectory.Samples(),
                           transformedTrajectory.Samples());
  TestSameForwardScalars(trajectory.Samples(), transformedTrajectory.Samples());
}

File: wpimath/src/test/java/org/wpilib/math/trajectory/TrajectoryTransformTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.trajectory;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import org.junit.jupiter.api.Test;
import org.wpilib.math.geometry.Pose2d;
import org.wpilib.math.geometry.Rotation2d;
import org.wpilib.math.geometry.Transform2d;
import org.wpilib.math.geometry.Translation2d;

class TrajectoryTransformTest {
  @Test
  void testTransformBy() {
    var trajectory =
        HolonomicTrajectoryGenerator.generate(
            Pose2d.ZERO, List.of(), new Pose2d(1, 1, Rotation2d.CCW_PI_2), 3, 1, 3, 1);

    var transformedTrajectory =
        trajectory.transformBy(
            new Transform2d(new Translation2d(1, 2), Rotation2d.fromDegrees(30)));

    // Test initial pose.
    assertEquals(
        new Pose2d(1, 2, Rotation2d.fromDegrees(30)), transformedTrajectory.sampleAt(0).pose);

    testSameShapedTrajectory(trajectory.samples, transformedTrajectory.samples);
    testSameForwardScalars(trajectory.samples, transformedTrajectory.samples);
  }

  @Test
  void testRelativeTo() {
    var trajectory =
        HolonomicTrajectoryGenerator.generate(
            new Pose2d(1, 2, Rotation2d.fromDegrees(30.0)),
            List.of(),
            new Pose2d(5, 7, Rotation2d.CCW_PI_2),
            3,
            1,
            3,
            1);

    var transformedTrajectory = trajectory.relativeTo(new Pose2d(1, 2, Rotation2d.fromDegrees(30)));

    // Test initial pose.
    assertEquals(Pose2d.ZERO, transformedTrajectory.sampleAt(0).pose);

    testSameShapedTrajectory(trajectory.samples, transformedTrajectory.samples);
    testSameForwardScalars(trajectory.samples, transformedTrajectory.samples);
  }

  // A rigid transform rotates both the heading and the field-relative velocity/acceleration by the
  // same amount, so the heading-relative forward scalars (and curvature) are invariant. This would
  // fail if transformBy/relativeTo rotated the pose but not the velocity/acceleration.
  void testSameForwardScalars(List<HolonomicSample> statesA, List<HolonomicSample> statesB) {
    assertEquals(statesA.size(), statesB.size());
    for (int i = 0; i < statesA.size(); i++) {
      assertEquals(statesA.get(i).velocity.vx, statesB.get(i).velocity.vx, 1e-9);
      assertEquals(statesA.get(i).velocity.vy, statesB.get(i).velocity.vy, 1e-9);
      assertEquals(statesA.get(i).velocity.omega, statesB.get(i).velocity.omega, 1e-9);
      assertEquals(statesA.get(i).acceleration.ax, statesB.get(i).acceleration.ax, 1e-9);
      assertEquals(statesA.get(i).acceleration.ay, statesB.get(i).acceleration.ay, 1e-9);
      assertEquals(statesA.get(i).acceleration.alpha, statesB.get(i).acceleration.alpha, 1e-9);
    }
  }

  void testSameShapedTrajectory(List<HolonomicSample> statesA, List<HolonomicSample> statesB) {
    for (int i = 0; i < statesA.size() - 1; i++) {
      var a1 = statesA.get(i).pose;
      var a2 = statesA.get(i + 1).pose;

      var b1 = statesB.get(i).pose;
      var b2 = statesB.get(i + 1).pose;

      var expectedRel = a2.relativeTo(a1);
      var actualRel = b2.relativeTo(b1);

      assertEquals(
          expectedRel,
          actualRel,
          String.format(
              "pose mismatch at index %d: expected %s, actual %s", i, expectedRel, actualRel));
    }
  }
}
